core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala

```diff
@@ -154,6 +154,113 @@ case class InputPSBTMap(elements: Vector[InputPSBTRecord])
   import org.bitcoins.core.psbt.InputPSBTRecord._
   import org.bitcoins.core.psbt.PSBTInputKeyId._
 
+  def nextRole(txIn: TransactionInput): PSBTRole = {
+    if (isFinalized) {
+      PSBTRole.ExtractorPSBTRole
+    } else {
+      (nonWitnessOrUnknownUTXOOpt, witnessUTXOOpt) match {
+        case (None, None) =>
+          PSBTRole.UpdaterPSBTRole
+        case (Some(_), None) | (None, Some(_)) | (Some(_), Some(_)) =>
+          val finalizeT = finalize(txIn)
+          finalizeT match {
+            case Failure(_) =>
+              val dummySigner = Sign.dummySign(ECPublicKey.freshPublicKey)
+              Try(toUTXOSigningInfo(txIn, dummySigner)) match {
+                case Failure(_) =>
+                  PSBTRole.SignerPSBTRole
+                case Success(_) =>
+                  PSBTRole.FinalizerPSBTRole
+              }
+            case Success(_) =>
+              PSBTRole.FinalizerPSBTRole
+          }
+      }
+    }
+  }
+
+  def prevOutOpt(vout: Int): Option[TransactionOutput] = {
+    witnessUTXOOpt match {
+      case Some(witnessUTXO) =>
+        Some(witnessUTXO.witnessUTXO)
+      case None =>
+        nonWitnessOrUnknownUTXOOpt match {
+          case Some(tx) =>
+            Some(tx.transactionSpent.outputs(vout))
+          case None => None
+        }
+    }
+  }
+
+  private def missingSigsFromScript(
+      spk: ScriptPubKey): Vector[Sha256Hash160Digest] = {
+    spk match {
+      case EmptyScriptPubKey | _: WitnessCommitment |
```

Comment:
Hmm.... should we throw in these cases? Is this logic used strictly for analyze or could it be used in other functionality too? 

I guess we probably shouldn't throw in the analyze case, but if we are hitting these script types in the sign case that is a problem.

Reply:
It shouldn't be a problem, if someone had a tx with an input that is an empty spk then it should tell them 0 sigs needed bc it doesn't need sigs. Likely if they do have an empty spk it is us testing something